drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/ModelBuilderImpl.java

```diff
@@ -303,4 +315,45 @@ public T getPackageSource(String packageName) {
     protected BuildContext createBuildContext() {
         return new CanonicalModelBuildContext();
     }
+
+    // Overridden, because exec-model doesn't add assets of included kbase to the packageDescr
+    @Override
+    protected void validateUniqueRuleNames(final PackageDescr packageDescr) {
+        super.validateUniqueRuleNames(packageDescr);
+
+        // check for duplicated rule names in included kbase
+        Map<String, InternalKieModule> includeModules = getBuildContext().getIncludeModules();
+        List<String> ruleNamesInIncludeKBases = new ArrayList<>();
+        for (Map.Entry<String, InternalKieModule> entry : includeModules.entrySet()) {
+            String kBaseName = entry.getKey();
+            InternalKieModule includeModule = entry.getValue();
+            if (!(includeModule instanceof CanonicalKieModule)) {
+                continue;
+            }
+            CanonicalKieModule canonicalKieModule = (CanonicalKieModule) includeModule;
+            InternalKieModule internalKieModule = canonicalKieModule.getInternalKieModule();
+            if (!(internalKieModule instanceof AbstractKieModule)) {
+                continue;
+            }
+            AbstractKieModule abstractKieModule = (AbstractKieModule) internalKieModule;
+            ReleaseId internalReleaseId = abstractKieModule.getReleaseId();
+            KieServices ks = KieServices.Factory.get();
+            KieContainer kieContainer = ks.newKieContainer(internalReleaseId);
+            KieBase kieBase = kieContainer.getKieBase(kBaseName);
+            // add rule names to ruleNamesInIncludeKBases
+            kieBase.getKiePackages().stream()
+                    .filter(kPkg -> kPkg.getName().equals(packageDescr.getNamespace()))
+                    .flatMap(kPkg -> kPkg.getRules().stream())
+                    .forEach(rule -> ruleNamesInIncludeKBases.add(rule.getName()));
```

Comment:
This may be a naive approach to collect rule names, but I cannot find a way to collect them directly from the include kbase's `KieModule`. 

There may be a case that we can get cached PackageDescr/RuleDescr of the included kbase.
```
    KnowledgeBuilder knowledgeBuilder = includeModule.getKnowledgeBuilderForKieBase(kBaseName);
    List<PackageDescr> includePackageDescrs = ((ModelBuilderImpl) knowledgeBuilder).getPackageDescrs(packageDescr.getNamespace());
```
but it's not possible in case of using kie-maven-plugin use case (`getKnowledgeBuilderForKieBase` returns `InternalKnowledgeBuilder$Empty`)

Reply:
This works, but in all honesty it seems also to be a huge amount of wasteful computational resources that moreover needs to happen again and again for each and every `PackageDescr` check. Just to figure out if we could find a better way to do this can you please point me to what does the non-exec model version? At the moment I don't understand why the interpreted version performs the names duplication check correctly while the exec model one doesn't. Also I wonder if we can make this more efficient by delaying this check at a later stage, maybe when the `KieBase` is created?

Reply:
> can you please point me to what does the non-exec model version?

https://github.com/kiegroup/drools/blob/7.x/drools-compiler/src/main/java/org/drools/compiler/kie/builder/impl/AbstractKieProject.java#L237-L239

non-exec model (`compileIncludedKieBases() returns true`) adds files of "included kbase" module into `assets`, so all of them will be parsed and compiled. Thus, it can validate the rule name duplication with `KnowledgeBuilderImpl.validateUniqueRuleNames`.

> it seems also to be a huge amount of wasteful computational resources

Yes. https://github.com/kiegroup/drools/pull/59 creates only KiePackage, not KieBase, so I hope it's relatively smaller.

> needs to happen again and again for each and every PackageDescr check

Thank you for pointing this out. I have moved `populateIncludedRuleNameMap()` out of packages loop in https://github.com/kiegroup/drools/pull/59, so executed only once.

> Also I wonder if we can make this more efficient by delaying this check at a later stage, maybe when the KieBase is created?

Sure, I'm going to do it as another PR. Note that it means the duplication cannot be detected at the time of kjar build by kie-maven-plugin, right?